src/dataflow/operator.rs

```diff
@@ -43,10 +36,84 @@ where
     fn destroy(&mut self) {}
 }
 
-/*****************************************************************************
- * Sink: receives data with type T                                           *
- *****************************************************************************/
+/*************************************************************************************************
+ * ReadOnlySink: receives data with type T, and enables message callbacks to execute in parallel *
+ * by only allowing append access in the message callbacks and writable access to the state in   *
+ * the watermark callbacks.                                                                      *
+ ************************************************************************************************/
+
+#[allow(unused_variables)]
+pub trait ReadOnlySink<S: ReadOnlyState<U, V>, T: Data, U, V>: Send + Sync {
+    fn run(&mut self, read_stream: &mut ReadStream<T>) {}
+
+    fn destroy(&mut self) {}
+
+    fn on_data(&self, ctx: &ReadOnlySinkContext<S, U, V>, data: &T);
+
+    fn on_watermark(&self, ctx: &mut ReadOnlySinkContext<S, U, V>);
+}
+
+pub struct ReadOnlySinkContext<'a, S: ReadOnlyState<T, U>, T, U> {
+    pub timestamp: Timestamp,
+    pub config: OperatorConfig,
+    pub state: &'a S,
```

Comment:
May want to hide these behind getters in a future PR.

Reply:
Agreed. Added.